### PR TITLE
Optimisation of count

### PR DESCRIPTION
**Pull request by gamdow**
Around 33% speed up over for-loop for large stencils (#460).

### CODE REVIEW

File: devito/symbolics/inspection.py [hunk 1]
```diff
@@ -7,18 +7,14 @@
...
 from devito.tools import flatten
 
-
 def count(exprs, query):
     """
     Return a mapper ``{(k, v)}`` where ``k`` is a sub-expression in ``exprs``
     matching ``query`` and ``v`` is the number of its occurrences.
     """
-    mapper = OrderedDict()
+    mapper = {}
```
Comment (FabioLuporini) [inline devito/symbolics/inspection.py]: Random guess.
Should this be a ``Counter`` too -- thus explaining why some tests are failing? The failures are due to capturing fewer common sub-expressions, which results in higher flop counts than what the tests are expecting. 

It's like with this solution, we get different ``hit``s here:
https://github.com/opesci/devito/blob/master/devito/dse/manipulation.py#L110
A ``hit`` is a candidate common sub-expression that gets replaced by a temporary

**Comment by gamdow** [comment]
You were right about the mapper needing to be a Counter.

**Comment by FabioLuporini** [comment]
Perhaps not an issue if you return a ``Counter``, rather than a ``dict``: if the primary concern was speed, returning a new ``dict`` will cause lots of key insertions, thus increasing the routine time cost? just a random thought.

Are the tests now passing?

Thanks

**Comment by FabioLuporini** [comment]
@gamdow just a flake8 issue (two consecutive blank lines at ``inspection.py:14``) and then this is probably good to go.

What's the ultimate speed improvement after switching to ``Counter``?

**Comment by FabioLuporini** [comment]
@gamdow I've updated the branch to latest ``master``, still keen to fix the remaining ``flake8`` issues? After that, we can land this

**Comment by gamdow** [comment]
What's CodeCov? I've never come across it before and the website is just a list of meaningless slogans.

**Comment by FabioLuporini** [comment]
@gamdow all good, thanks

**Pull request merged**